### ISSUE TITLE
Logging error in SFTP copy with `Sequence[PurePath]`

### ISSUE THREAD

**Issue by eyalgolan1337** [ronf/asyncssh#668]
When copying with SFTP, a log is generated containing `srcpaths`. If this is a list, the `_text` function attempts to call `str.join` with it, which fails for `PurePath`.

The `_text` function also assumes that all arguments are of the same type, but `put`/`get`/`copy` accept a mix of `str`/`bytes`/`PurePath`. (Maybe `_text` should be recursive?)

Additionally, these functions are typed as accepting a `Sequence` but actually only accept tuples and lists. On the other hand, they accept single values but are typed as only accepting sequences. The type of `srcpaths` should be `_SFTPPaths` instead of `Sequence[_SFTPPath]`.

**Comment by ronf** [comment]
Thanks for the report. I'm able to reproduce the issue here.

In terms of data types, the intention is to for these functions to support any sequence. If you look in`_begin_copy()`, it converts a tuple to a list, but that's only because the logging code only supports tuples for logging host/port values. Otherwise, `_begin_copy()` checks if the input is bytes/str/PurePath and turns those into single-element lists but expects any other data type to be something it can run a "for" loop on and (putting aside logging issues), that allows for things like a generator or other iterable to be passed in.

That said, you're right that the type should be _SFTPPaths, since it allows either a singleton bytes/str/PurePath or a sequence of them.

As for the logging, you're right that it currently only supports lists which are either all bytes or all str, and it doesn't properly support including PurePath in a list (though it works fine as a singleton).

I'll have to think a bit about how to fix the logging. I'd prefer not to go fully recursive, as the logging function was only intended to support a single level of list and not nested lists. I'll probably break it into two functions, one to convert single elements and another to handle the case where it is a list.

There's another issue as well, or rather there will be if I fix the above -- I won't be able to support srcpaths being a generator function if I end up iterating over it in the logging code. In that case, I'd need to convert it to a list up front so both the logging code and the main code in `_begin_copy()` can iterate over the items. So, that'll require some rearranging, but it shouldn't be too difficult.

I'll let you know when I have something to test.

**Comment by ronf** [comment]
Ok - here's a first cut at a fix:

```patch
diff --git a/asyncssh/logging.py b/asyncssh/logging.py
index 828d85c..d28eb3b 100644
--- a/asyncssh/logging.py
+++ b/asyncssh/logging.py
@@ -61,14 +61,28 @@ class SSHLogger(logging.LoggerAdapter):
     def log(self, level: int, msg: object, *args, **kwargs) -> None:
         """Log a message to the underlying logger"""

-        def _text(arg: _LogArg) -> str:
+        def _item_text(item: _LogArg) -> str:
+            """Convert a list item to text"""
+
+            if isinstance(item, bytes):
+                result = item.decode('utf-8', errors='replace')
+
+                if not result.isprintable():
+                    result = repr(result)[1:-1]
+            elif not isinstance(item, str):
+                result = str(item)
+            else:
+                result = item
+
+            return result
+
+        def _text(arg: _LogArg) -> _LogArg:
             """Convert a log argument to text"""

+            result: _LogArg
+
             if isinstance(arg, list):
-                if arg and isinstance(arg[0], bytes):
-                    result = b','.join(arg).decode('utf-8', errors='replace')
-                else:
-                    result = ','.join(arg)
+                result = ','.join(_item_text(item) for item in arg)
             elif isinstance(arg, tuple):
                 host, port = arg

@@ -76,14 +90,10 @@ class SSHLogger(logging.LoggerAdapter):
                     result = '%s, port %d' % (host, port) if port else host
                 else:
                     result = 'port %d' % port if port else 'dynamic port'
+            elif isinstance(arg, bytes):
+                result = _item_text(arg)
             else:
-                result = cast(str, arg)
-
-            if isinstance(result, bytes):
-                result = result.decode('ascii', errors='backslashreplace')
-
-                if not result.isprintable():
-                    result = repr(result)[1:-1]
+                result = arg

             return result

diff --git a/asyncssh/sftp.py b/asyncssh/sftp.py
index 3157096..d3498fc 100644
--- a/asyncssh/sftp.py
+++ b/asyncssh/sftp.py
@@ -3679,8 +3679,7 @@ class SFTPClient:
                 raise

     async def _begin_copy(self, srcfs: _SFTPFSProtocol, dstfs: _SFTPFSProtocol,
-                          srcpaths: Sequence[_SFTPPath],
-                          dstpath: Optional[_SFTPPath],
+                          srcpaths: _SFTPPaths, dstpath: Optional[_SFTPPath],
                           copy_type: str, expand_glob: bool, preserve: bool,
                           recurse: bool, follow_symlinks: bool,
                           block_size: int, max_requests: int,
@@ -3688,15 +3687,14 @@ class SFTPClient:
                           error_handler: SFTPErrorHandler) -> None:
         """Begin a new file upload, download, or copy"""

-        if isinstance(srcpaths, tuple):
+        if isinstance(srcpaths, (bytes, str, PurePath)):
+            srcpaths = [srcpaths]
+        elif not isinstance(srcpaths, list):
             srcpaths = list(srcpaths)

         self.logger.info('Starting SFTP %s of %s to %s',
                          copy_type, srcpaths, dstpath)

-        if isinstance(srcpaths, (bytes, str, PurePath)):
-            srcpaths = [srcpaths]
-
         srcnames: List[SFTPName] = []

         if expand_glob:
@@ -3741,7 +3739,7 @@ class SFTPClient:
                              preserve, recurse, follow_symlinks, block_size,
                              max_requests, progress_handler, error_handler)

-    async def get(self, remotepaths: Sequence[_SFTPPath],
+    async def get(self, remotepaths: _SFTPPaths,
                   localpath: Optional[_SFTPPath] = None, *,
                   preserve: bool = False, recurse: bool = False,
                   follow_symlinks: bool = False,
@@ -3846,7 +3844,7 @@ class SFTPClient:
                                block_size, max_requests, progress_handler,
                                error_handler)

-    async def put(self, localpaths: Sequence[_SFTPPath],
+    async def put(self, localpaths: _SFTPPaths,
                   remotepath: Optional[_SFTPPath] = None, *,
                   preserve: bool = False, recurse: bool = False,
                   follow_symlinks: bool = False,
@@ -3951,7 +3949,7 @@ class SFTPClient:
                                block_size, max_requests, progress_handler,
                                error_handler)

-    async def copy(self, srcpaths: Sequence[_SFTPPath],
+    async def copy(self, srcpaths: _SFTPPaths,
                    dstpath: Optional[_SFTPPath] = None, *,
                    preserve: bool = False, recurse: bool = False,
                    follow_symlinks: bool = False,
@@ -4056,7 +4054,7 @@ class SFTPClient:
                                block_size, max_requests, progress_handler,
                                error_handler)

-    async def mget(self, remotepaths: Sequence[_SFTPPath],
+    async def mget(self, remotepaths: _SFTPPaths,
                    localpath: Optional[_SFTPPath] = None, *,
                    preserve: bool = False, recurse: bool = False,
                    follow_symlinks: bool = False,
@@ -4080,7 +4078,7 @@ class SFTPClient:
                                block_size, max_requests, progress_handler,
                                error_handler)

-    async def mput(self, localpaths: Sequence[_SFTPPath],
+    async def mput(self, localpaths: _SFTPPaths,
                    remotepath: Optional[_SFTPPath] = None, *,
                    preserve: bool = False, recurse: bool = False,
                    follow_symlinks: bool = False,
@@ -4104,7 +4102,7 @@ class SFTPClient:
                                block_size, max_requests, progress_handler,
                                error_handler)

-    async def mcopy(self, srcpaths: Sequence[_SFTPPath],
+    async def mcopy(self, srcpaths: _SFTPPaths,
                     dstpath: Optional[_SFTPPath] = None, *,
                     preserve: bool = False, recurse: bool = False,
                     follow_symlinks: bool = False,
```

**Comment by ronf** [comment]
This change is now available in the "develop" branch as commit b21e758. Let me know if you run into any issues with it.

**Comment by ronf** [comment]
This change is now available in AsyncSSH 2.16.0.
